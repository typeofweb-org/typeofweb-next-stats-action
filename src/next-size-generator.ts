import type { SizeUnit, SnapshotSizeEntry } from './types';
import { prettyBytesInverse } from './utils';

type Group = {
  readonly treeViewPresentation: string;
  readonly subTree?: string;
  readonly fileType?: string;
  readonly pageUrl: string;
  readonly sizeFormatted: string;
  readonly sizeUnit: SizeUnit;
  // eslint-disable-next-line functional/prefer-readonly-type
  children: Group[];
};

const pageRegex = /(?<treeViewPresentation>┌|├|└)\s+((?<subTree>┌|├|└)\s+)?((?<fileType>λ|○|●)\s+)?(?<pageUrl>[^\s]+)\s+(?<sizeFormatted>[0-9.]+)\s+(?<sizeUnit>\w+)/gm;

export function getNextPagesSize(consoleOutput: string) {
  // eslint-disable-next-line functional/prefer-readonly-type
  const jsChunks: number[] = [];
  // eslint-disable-next-line functional/prefer-readonly-type
  const cssChunks: number[] = [];

  const rows = (Array.from(consoleOutput.matchAll(pageRegex)) as unknown) as ReadonlyArray<{
    readonly groups: Group;
  }>;

  const groupedRows = rows
    .map((r) => r.groups)
    // eslint-disable-next-line functional/prefer-readonly-type
    .reduce<Group[]>((acc, group, index) => {
      if (group.pageUrl === '/_app') {
        // skip because it's included later on as /static/pages/_app so this is a duplicate
        return acc;
      }
      if (group.subTree && index > 0) {
        // eslint-disable-next-line functional/no-loop-statement
        for (let i = acc.length - 1; i >= 0; --i) {
          if (!acc[i]!.subTree) {
            acc[i]!.children = acc[i]!.children || [];
            acc[i]!.children.push(group);
            return acc;
          }
        }
      }
      acc.push({ ...group, children: [] });
      return acc;
    }, []);

  const entries = groupedRows.map((row) => {
    const { pageUrl, sizeFormatted, sizeUnit, children } = row;
    let snapshotId = `${pageUrl}`;

    if (pageUrl.startsWith('css/')) {
      cssChunks.push(prettyBytesInverse(sizeFormatted, sizeUnit));
      return null;
    } else if (pageUrl.includes('pages/_app.')) {
      snapshotId = 'shared:_app.js';
    } else if (/^runtime\/main\.(.+)\.js$/.test(pageUrl)) {
      snapshotId = 'shared:runtime/main';
    } else if (/^runtime\/webpack\.(.+)\.js$/.test(pageUrl)) {
      snapshotId = 'shared:runtime/webpack';
    } else if (/^chunks\/commons\.(.+)\.js$/.test(pageUrl)) {
      snapshotId = 'shared:chunk/commons';
    } else if (/^chunks\/MarkdownText\.(.+)\.js$/.test(pageUrl)) {
      snapshotId = 'shared:MarkdownText';
    } else if (/^chunks\/framework\.(.+)\.js$/.test(pageUrl)) {
      snapshotId = 'shared:chunk/framework';
    } else if (/^chunks\/(.*)\.js$/.test(pageUrl)) {
      // shared chunks are unnamed and only have a hash
      // we just track their count and summed size
      jsChunks.push(prettyBytesInverse(sizeFormatted, sizeUnit));
      // and not each chunk individually
      return null;
    }

    const childrenSizeSum = children.reduce((acc, group) => {
      return acc + prettyBytesInverse(group.sizeFormatted, group.sizeUnit);
    }, 0);

    return [
      snapshotId,
      {
        parsed: prettyBytesInverse(sizeFormatted, sizeUnit),
        childrenSize: childrenSizeSum,
        children: children.reduce((acc, c) => {
          const key = c.pageUrl.replace(/\/[a-z0-9]{20}/, '/[hash]');
          acc[key] = acc[key] || { parsed: 0 };
          acc[key]!.parsed += prettyBytesInverse(c.sizeFormatted, c.sizeUnit);
          return acc;
          // eslint-disable-next-line functional/prefer-readonly-type
        }, {} as Record<string, { parsed: number }>),
      },
    ] as SnapshotSizeEntry;
  });

  const nonEmptyEntries = entries.filter((entry): entry is SnapshotSizeEntry => entry != null);

  nonEmptyEntries.push([
    'shared:js',
    {
      parsed: jsChunks.reduce((sum, size) => sum + size, 0),
      count: jsChunks.length,
      childrenSize: 0,
      children: {},
    },
  ]);
  nonEmptyEntries.push([
    'shared:css',
    {
      parsed: cssChunks.reduce((sum, size) => sum + size, 0),
      count: cssChunks.length,
      childrenSize: 0,
      children: {},
    },
  ]);

  return Object.fromEntries(nonEmptyEntries);
}
