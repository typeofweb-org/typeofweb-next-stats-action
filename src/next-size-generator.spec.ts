import { getNextPagesSize } from './next-size-generator';
import 'jest-extended';

describe('getNextPagesSize', () => {
  it('should work', () => {
    const consoleOutput = `
    Page                                                            Size     First Load JS
    ┌   /_app                                                       0 B             105 kB
    ├ λ /_err                                                       1.79 kB         107 kB
    ├ ● /[displayStyle]                                             2.53 kB         147 kB
    ├   ├ /grid
    ├   └ /list
    ├ ● /[displayStyle]/[cursor]                                    2.51 kB         147 kB
    ├   ├ /grid/xlaydiPAk6Ed3Kfkq96qeOp5hPc3yQ0QlwVL%2F9ZyjBo%3D
    ├   ├ /grid/RKPcckbiPMH2F72nit2U6Syg996wfWht96e7yXFsaSc%3D
    ├   ├ /grid/AsBI640yhccb%2FqOlAVZOWQNX3h3p%2FklVnHkmEQtnaPg%3D
    ├   └ [+7 more paths]
    ├ ○ /404                                                        3.82 kB         109 kB
    ├ ○ /admin                                                      2.81 kB         111 kB
    ├   └ css/0b06d3dc729f6afe3013.css                              2.61 kB
    ├ λ /api                                                        0 B             105 kB
    ├ λ /api/auth/[...nextauth]                                     0 B             105 kB
    ├ λ /api/blogs                                                  0 B             105 kB
    ├ λ /api/blogs/[blogId]                                         0 B             105 kB
    ├ λ /api/content-creator                                        0 B             105 kB
    ├ λ /api/feed                                                   0 B             105 kB
    ├ λ /api/update-blogs                                           0 B             105 kB
    ├ λ /api/update-feed                                            0 B             105 kB
    ├ ● /artykuly/[slug]                                            1.15 kB         145 kB
    ├   └ css/d8b374e26cdf201c0bbe.css                              3.05 kB
    ├   ├ /artykuly/parcel-bundler-czyli-webpack-bez-konfiguracji
    ├   ├ /artykuly/5-bibliotek-js'a-ktore-zawladna-rynkiem-w-2021
    ├   ├ /artykuly/5-najlepszych-ksiazek-2020
    ├   └ [+27 more paths]
    ├ ○ /o-serwisie                                                 2.26 kB         111 kB
    ├   └ css/5c4ba465e3ea75287602.css                              2.96 kB
    └ ○ /zglos-serwis                                               4.66 kB         113 kB
        └ css/7bb9b00fd266464361e6.css                              3.18 kB
    + First Load JS shared by all                                   105 kB
      ├ chunks/8c43c7dba21b70476fcfc932bb90be731e9df189.f9f6ba.js   13.8 kB
      ├ chunks/8d2a18b53678a0436e3e4bc986ccc9734d9e4be5.dfdb83.js   24 kB
      ├ chunks/a9f427f82024d0e57d1fcf0eee6531ceed07b6c4.30b228.js   14 kB
      ├ chunks/f00a516035ddd5a7692aff084b0e21e2ba566403.e2c5f9.js   3.3 kB
      ├ chunks/framework.c4c935.js                                  42 kB
      ├ chunks/main.b755b8.js                                       6.3 kB
      ├ chunks/pages/_app.1631bc.js                                 231 B
      ├ chunks/webpack.848c30.js                                    1.56 kB
      └ css/ec77e3ea886b8ea8cf98.css                                2 kB
    
    λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
    ○  (Static)  automatically rendered as static HTML (uses no initial props)
    ●  (SSG)     automatically generated as static HTML + JSON (uses getStaticProps)
       (ISR)     incremental static regeneration (uses revalidate in getStaticProps)
    
    ✨  Done in 27.57s.
    `.trim();

    const result = getNextPagesSize(consoleOutput);

    const expectedKeys = [
      '/_err',
      '/[displayStyle]',
      '/[displayStyle]/[cursor]',
      '/404',
      '/admin',
      '/artykuly/[slug]',
      '/o-serwisie',
      '/zglos-serwis',
      'shared:chunk/framework',
      'shared:chunk/main',
      'shared:_app.js',
      'shared:chunk/webpack',
      'shared:js',
      'shared:css',
    ];
    console.log(result);

    expect(result).toContainKeys(expectedKeys);
    expect(result).toMatchSnapshot();
  });
});
