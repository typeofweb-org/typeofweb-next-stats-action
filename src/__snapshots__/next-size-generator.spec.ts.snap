// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getNextPagesSize should work 1`] = `
Object {
  "/404": Object {
    "children": Object {},
    "childrenSize": 0,
    "parsed": 3820,
    "renderType": "○",
  },
  "/[displayStyle]": Object {
    "children": Object {},
    "childrenSize": 0,
    "parsed": 2530,
    "renderType": "●",
  },
  "/[displayStyle]/[cursor]": Object {
    "children": Object {},
    "childrenSize": 0,
    "parsed": 2510,
    "renderType": "●",
  },
  "/_err": Object {
    "children": Object {},
    "childrenSize": 0,
    "parsed": 1790,
    "renderType": "λ",
  },
  "/admin": Object {
    "children": Object {
      "css/[hash].css": Object {
        "parsed": 2610,
      },
    },
    "childrenSize": 2610,
    "parsed": 2810,
    "renderType": "○",
  },
  "/artykuly/[slug]": Object {
    "children": Object {
      "css/[hash].css": Object {
        "parsed": 3050,
      },
    },
    "childrenSize": 3050,
    "parsed": 1150,
    "renderType": "●",
  },
  "/o-serwisie": Object {
    "children": Object {
      "css/[hash].css": Object {
        "parsed": 2960,
      },
    },
    "childrenSize": 2960,
    "parsed": 2260,
    "renderType": "○",
  },
  "/zglos-serwis": Object {
    "children": Object {},
    "childrenSize": 0,
    "parsed": 4660,
    "renderType": "○",
  },
  "shared:_app.js": Object {
    "children": Object {},
    "childrenSize": 0,
    "parsed": 231,
    "renderType": "",
  },
  "shared:chunk/framework": Object {
    "children": Object {},
    "childrenSize": 0,
    "parsed": 42000,
    "renderType": "",
  },
  "shared:chunk/main": Object {
    "children": Object {},
    "childrenSize": 0,
    "parsed": 6300,
    "renderType": "",
  },
  "shared:chunk/webpack": Object {
    "children": Object {},
    "childrenSize": 0,
    "parsed": 1560,
    "renderType": "",
  },
  "shared:css": Object {
    "children": Object {},
    "childrenSize": 0,
    "count": 2,
    "parsed": 5180,
  },
  "shared:js": Object {
    "children": Object {},
    "childrenSize": 0,
    "count": 4,
    "parsed": 55100,
  },
}
`;
